common/src/chain/block/block_body/merkle_proxy.rs

```diff
@@ -0,0 +1,171 @@
+    merkle_tools::MerkleHasher,
+    BlockBody, BlockMerkleTreeError,
+};
+
+#[must_use]
+pub struct BlockBodyMerkleProxy {
+    witness_tree: WrappedWitnessMerkleTree,
+    tree: WrappedMerkleTree,
+}
+
+impl BlockBodyMerkleProxy {
+    pub fn new(body: &BlockBody) -> Result<Self, BlockMerkleTreeError> {
+        Ok(Self {
+            witness_tree: calculate_witness_merkle_tree(body)?.into(),
+            tree: calculate_tx_merkle_tree(body)?.into(),
+        })
+    }
+
+    pub fn witness_merkle_tree(&self) -> &WrappedWitnessMerkleTree {
+        &self.witness_tree
+    }
+
+    pub fn merkle_tree(&self) -> &WrappedMerkleTree {
+        &self.tree
+    }
+}
+
+/// This struct wrapper, and the other one, are an attempt to create minimal
+/// type safety to avoid confusing the two merkle trees, with and without
+/// the witness.
+#[must_use]
+pub struct WrappedWitnessMerkleTree {
+    witness_merkle_tree: MerkleTree<H256, MerkleHasher>,
+}
+
+impl From<MerkleTree<H256, MerkleHasher>> for WrappedWitnessMerkleTree {
+    fn from(witness_merkle_tree: MerkleTree<H256, MerkleHasher>) -> Self {
+        Self {
+            witness_merkle_tree,
+        }
+    }
+}
+
+impl WrappedWitnessMerkleTree {
+    pub fn root(&self) -> H256 {
+        self.witness_merkle_tree.root()
+    }
+
+    pub fn raw_tree(&self) -> &MerkleTree<H256, MerkleHasher> {
+        &self.witness_merkle_tree
+    }
+
+    pub fn block_reward_witness_leaf(&self) -> H256 {
+        *self
+            .witness_merkle_tree
+            .node_from_bottom(0, 0)
+            .expect("Block reward leaf must exist")
+            .hash()
+    }
+
+    pub fn transaction_witness_leaf(&self, index: usize) -> H256 {
+        *self
+            .witness_merkle_tree
+            .node_from_bottom(0, index as u32 + 1) // +1 because of block reward leaf
+            .expect("Transaction witness leaf must exist")
+            .hash()
+    }
+
+    pub fn block_reward_inclusion_proof(
+        &self,
+    ) -> Result<SingleProofHashes<H256, MerkleHasher>, BlockMerkleTreeError> {
+        // Block reward has index 0 in the block merkle tree
+        let proof = SingleProofNodes::from_tree_leaf(&self.witness_merkle_tree, 0)?;
+
+        Ok(proof.into_values())
+    }
+
+    pub fn transaction_witness_inclusion_proof(
+        &self,
+        index_in_block: u32,
+    ) -> Result<SingleProofHashes<H256, MerkleHasher>, BlockMerkleTreeError> {
+        // We add 1 to the index_in_block because the block reward is the first element in the block merkle tree
+        let proof =
+            SingleProofNodes::from_tree_leaf(&self.witness_merkle_tree, index_in_block + 1)?;
+
+        Ok(proof.into_values())
+    }
+}
+
+/// This struct wrapper, and the other one, are an attempt to create minimal
+/// type safety to avoid confusing the two merkle trees, with and without
+/// the witness.
+#[must_use]
+pub struct WrappedMerkleTree {
+    merkle_tree: MerkleTree<H256, MerkleHasher>,
+}
```

Comment:
Looks like this struct and `WrappedWitnessMerkleTree` are identical, including methods. How about having one generic structure `MerkleTreeOf<Something>` where `Something` is just a type tag, stored in `PhantomData`, specifying what sort of data the tree refers to. Then we can have two distinct types `MerkleTreeOf<Transaction>` and `MerkleTreeOf<Witness>` which prevents from them being mixed up. Methods can be implemented in one `impl` block generic over the tag type.

Reply:
That's reasonable. But not with `Witness` and `Transaction`. I'll create custom tag-structs.

Reply:
OK... now I remember why I didn't do that... I wanted the method names to be different so that a copy-paste error doesn't happen...

Reply:
> OK... now I remember why I didn't do that... I wanted the method names to be different so that a copy-paste error doesn't happen...

You can sort of do it by introducing separate `impl` blocks for `MerkleTreeOf<Witness>` and `MerkleTreeOf<Transaction>` (or whatever names you end up choosing). That is in addition to the generic `impl` blocks for methods that are shared. You can use the generic block for shared implementation of the methods with different names if needed, just keep it private.

Reply:
Alrighty. I'll give that a shot. Cheers!

Reply:
Done.